torch_xla/experimental/scan.py

```diff
@@ -82,14 +132,514 @@ def scan(fn, init, xs):
+  and another function that will add back the `None` values when given a
+  sequence of the same structure.
+  """
+  filtered = [v for v in s if v is not None]
+  none_mask = [v is None for v in s]
+
+  def add_back_nones(s_filtered):
+    res = []
+    idx_filtered = 0
+    for is_none in none_mask:
+      if is_none:
+        res.append(None)
+      else:
+        res.append(s_filtered[idx_filtered])
+        idx_filtered += 1
+    return res
+
+  return filtered, add_back_nones
+
+
+def dynamic_update_slice(ys: xb.Op, y: xb.Op, idx: xb.Op) -> xb.Op:
+  # See https://openxla.org/xla/operation_semantics#dynamicupdateslice.
+  y = y.broadcast([1])
+  indices = [idx]
+  for _ in range(ys.shape().rank - 1):
+    indices.append(idx.zeros_like())
+  return ys.dynamic_update_slice(y, indices)
+
+
+def dynamic_slice(xs: xb.Op, idx: xb.Op) -> xb.Op:
+  indices = [idx]
+  for _ in range(xs.shape().rank - 1):
+    indices.append(idx.zeros_like())
+  slice_shape = list(xs.shape().sizes)
+  slice_shape[0] = 1
+  sliced = xs.dynamic_slice(indices, slice_shape)
+  shape = list(xs.shape().sizes)
+  shape = shape[1:]
+  return sliced.reshape(shape)
+
+
+class Builder:
+
+  def __init__(self, name: str):
+    self._builder = xb.create_builder(name)
+    self._params = []
+    self._param_tensors = []
+
+  def add_param(self, val: torch.Tensor):
+    idx = len(self._params)
+    param = xb.mkparam(self._builder, idx, xb.tensor_shape(val))
+    self._params.append(param)
+    self._param_tensors.append(val)
+    return idx
+
+  def params(self) -> Tuple[xb.Op, ...]:
+    return tuple(self._params)
+
+  def param_tensors(self) -> Tuple[torch.Tensor, ...]:
+    return tuple(self._param_tensors)
+
+  def num_params(self) -> int:
+    return len(self._params)
+
+
+def _scan_impl_flat(fn,
+                    init: Sequence[torch.Tensor],
+                    xs: Sequence[torch.Tensor],
+                    reverse: bool = False):
+  """Forward logic of scan without gradient tracking. `fn` operates on
+  two flat list of tensors. `init` and `xs` are also flat lists of tensors. None
+  of the tensors will be `None`.
+  
+  See the `Scan` class which implements an autograd `Function` and builds
+  autograd support on top of `_scan_impl`.
+  
+  ## Handling of random numbers
+
+  When `fn` generates random numbers (e.g. it uses a dropout layer), we need to
+  ensure that each iteration of `fn` within the scan yields different random
+  numbers, despite running the same HLO operations. JAX requires the user to
+  explicitly fork the RNG state and pass it to `fn`. In PyTorch, the RNG state
+  is an implicit global variable. Therefore, we take a slightly different
+  approach:
+
+  - Identify usage of RNG state via `_get_tensors_xla_device_data_node`.
+  - Create N different copies of the RNG state contained in a tensor.
+  - While building the `While` op body, index into the RNG state tensor at the
+    current iteration and provide that seed value to `fn`.
```

Comment:
Ah this is because if you use aot_autograd to capture graph the random seed is not part of the graph...